internal/config/funcs/jsonnet.go

```diff
@@ -0,0 +1,221 @@
+package funcs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-jsonnet"
+	"github.com/hashicorp/go-multierror"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+func Jsonnet() map[string]function.Function {
+	return map[string]function.Function{
+		"jsonnetdir":  JsonnetDirFunc,
+		"jsonnetfile": JsonnetFileFunc,
+	}
+}
+
+func jsonnetVM(opts cty.Value) (*jsonnet.VM, error) {
+	vm := jsonnet.MakeVM()
+
+	// Get our options. If it isn't a valid type that we're looking for
+	// then just return the VM as-is.
+	optsT := opts.Type()
+	if opts.IsNull() {
+		return vm, nil
+	}
+	if !optsT.IsObjectType() && !optsT.IsMapType() {
+		return vm, nil
+	}
+
+	// Jsonnet has FOUR ways to parameterize a file, all with string key/values.
+	// This function abstracts setting it so we can set all four methods.
+	var erracc error
+	setter := func(k string, cb func(string, string)) {
+		val := attr(opts, k)
+		if !val.IsNull() {
+			if !val.CanIterateElements() {
+				erracc = multierror.Append(erracc, fmt.Errorf(
+					"Option %q must be an object or map", k))
+				return
+			}
+
+			for optK, v := range val.AsValueMap() {
+				if v.Type() != cty.String {
+					erracc = multierror.Append(erracc, fmt.Errorf(
+						"Option %q values must be string types", k))
+					return
+				}
+
+				cb(optK, v.AsString())
+			}
+
+		}
+	}
+	setter("ext_vars", vm.ExtVar)
+	setter("ext_code", vm.ExtCode)
+	setter("tla_vars", vm.TLAVar)
+	setter("tla_code", vm.TLACode)
+
+	if erracc != nil {
+		return nil, erracc
+	}
+
+	return vm, nil
+}
+
+func attr(v cty.Value, k string) cty.Value {
+	t := v.Type()
+	if t.IsMapType() {
+		return v.Index(cty.StringVal(k))
+	} else if t.HasAttribute(k) {
+		return v.GetAttr(k)
+	}
+
+	return cty.NullVal(cty.String)
+}
+
+// JsonnetDirFunc constructs a function that converts a directory of
+// jsonnet files into standard JSON files with the same name but a "json"
+// extension instead of the "jsonnet" extension. The converted files are
+// stored in a temporary directory that is returned; the original files are
+// untouched.
+var JsonnetDirFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "dir",
+			Type: cty.String,
+		},
+		{
+			Name: "options",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.String),
+	Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+		td, err := ioutil.TempDir("", "waypoint")
```

Comment:
What deletes the tempdir?

Reply:
Nobody, this is behaving the same way as `template*` functions at the moment where we dangle the temp dirs. The hope is it goes into a system temp dir that gets cleared at some point currently (mine does on both macOS and Linux, at least). I think in the future we can try some process-wide cleanup if you wanted but I think that should be a separate PR.